File: packages/lua/stack.lua
local stack = {}

local Stack = {}
Stack.__index = Stack

function stack.New()
    return setmetatable({elements = {}}, Stack)
end

function Stack:Push(value)
    table.insert(self.elements, value)
end

function Stack:Pop()
    if self:IsEmpty() then return nil, "stack is empty" end
    return table.remove(self.elements)
end

function Stack:Peek()
    if self:IsEmpty() then return nil, "stack is empty" end
    return self.elements[#self.elements]
end

function Stack:IsEmpty()
    return #self.elements == 0
end

function Stack:Size()
    return #self.elements
end

function stack.IsValidParentheses(s)
    local st = stack.New()
    local pairs = { [')'] = '(', ['}'] = '{', [']'] = '[' }
    for i = 1, #s do
        local char = s:sub(i, i)
        if char == '(' or char == '{' or char == '[' then
            st:Push(char)
        elseif pairs[char] then
            if st:IsEmpty() or st:Pop() ~= pairs[char] then
                return false
            end
        end
    end
    return st:IsEmpty()
end

return stack

File: packages/rust/src/stack.rs
pub struct Stack<T> {
    elements: Vec<T>,
}

impl<T> Stack<T> {
    pub fn new() -> Self {
        Stack { elements: Vec::new() }
    }

    pub fn push(&mut self, item: T) {
        self.elements.push(item);
    }

    pub fn pop(&mut self) -> Option<T> {
        self.elements.pop()
    }

    pub fn is_empty(&self) -> bool {
        self.elements.is_empty()
    }
}

pub fn is_valid_parentheses(s: &str) -> bool {
    let mut stack = Stack::new();
    for c in s.chars() {
        match c {
            '(' | '{' | '[' => stack.push(c),
            ')' => if stack.pop() != Some('(') { return false; },
            '}' => if stack.pop() != Some('{') { return false; },
            ']' => if stack.pop() != Some('[') { return false; },
            _ => (),
        }
    }
    stack.is_empty()
}
